### DISCUSSION TITLE
V1.0.0 Testing

### DISCUSSION THREAD

**Discussion by Colt-H** [ottowayi/pycomm3#113] [General]
Is there a standard for how we should test devices, or are you just looking for anything?

I have access to L6/7/8 series controllogix and 1768/69/5069 series compact logix of any firmware after 17, along with tons of other EIP devices, all in a dedicated testing environment.

If you can let me know what you want to see in tests, I can knock these out as I get time at work.

**Comment by ottowayi** [comment]
That would be extremely helpful, thanks.  If you want to do the full unit tests, in the tests directory is a L5X program export that has all the tags needed.   The tests can be run with tox by `tox -e user` , if you aren't familiar with tox I can help you set it up.   That is going above and beyond though, simply reading/writing a few tags in each PLC would be more than enough.  If you already have scripts running, testing them against the new version would be good too.  I'm mostly trying to make sure that I didn't break anything.   The first main thing I can't test with my devices is the use of instance IDs in the read/write requests, that requires a V21+ PLC.  The other is the large connection/extended forward open that requires a V20+ and EN2T+ (or a newer CompactLogix).   Those features are enabled automatically on connect, so just reading/writing a couple tags and sending me the logs would be enough to confirm they work.  The new version provides a helper method to enable basic logging:

```
from pycomm3.logger import configure_default_logger, LOG_VERBOSE
configure_default_logger(level=LOG_VERBOSE)
```

Logging to a file would be the simplest though using:

```
from pycomm3.logger import configure_default_logger, LOG_VERBOSE, logger
import logging
configure_default_logger(level=LOG_VERBOSE)
handler = logging.FileHandler('pycomm3-test.log', encoding='utf-8')
logger.addHandler(handler)
```

And now I think I'm going to add the file logging option to the `configure_default_logger` method.

**Comment by Colt-H** [comment]
Okay I'll see what all I can do. For time's sake I might just run the full test on a couple higher FW versions, and some basic read/writes on others. It might take me a few weeks to get through everything I have, but I'll prioritize what you listed. I can definitely play around with it and log everything. I need to practice some coding anyway.

Do you want log files posted in the issues thread?

-------- Original Message --------

**Comment by Colt-H** [comment]
I'm not entirely sure what I'm doing wrong here, but I can't get reads or writes to work using V1.0.0. PLC is a 1756-L71 V24.11. Running the following code:

```
from pycomm3 import LogixDriver
from pycomm3.logger import configure_default_logger, LOG_VERBOSE

configure_default_logger(level=LOG_VERBOSE)
def read_single():
    with LogixDriver('10.64.138.13/0') as plc:
        return plc.read('TestDINT')

tag = read_single()
print(tag)
```
will only return ``True`` and nothing else. The log output shows it opening the connection, sending and receiving packets correctly, and closing successfully. This code (without logging) works successfully with 0.12.0.

I can successfully read the PLC Data using 

```
def read_PLC():
    with LogixDriver('10.64.138.13/0') as plc:
        return plc

tag = read_PLC()

print(tag)
```

![image](https://user-images.githubusercontent.com/65498634/108606699-84c2ba00-7389-11eb-961e-13283dd5b6f2.png)

**Reply by ottowayi** [reply]
Yeah that is a bug I just pushed the fix for.  I'm assuming you installed from github and not PyPI right?  The version on PyPI is missing some changes that are in the `develop` branch.  No major functionality, but some minor tweaks and logging.  If you re-install from `develop` now it should be fixed.

**Reply by Colt-H** [reply]
Yeah I cloned directly from github. Let me clone again and give it a go.

**Reply by Colt-H** [reply]
Okay that took care of it. 

Was the log below enough to confirm Instance IDs for read/writes?

**Reply by ottowayi** [reply]
Okay good, I was hoping that was it.  I usually test my changes before pushing them, but I forgot this time.  Unfortunately, that bug prevented the read method from even being called, so it's not in the log.

**Reply by ottowayi** [reply]
Since you're running from the repo, there is a `filename` option for that logging config method now.  You can use that and just attach the file if that makes things easier.

**Reply by Colt-H** [reply]
Okay cool, I'll do that.

**Reply by Colt-H** [reply]
This was weird. So I'm currently running V1.0.0 cloned directly from the develop branch of github. 

I enabled filename logging and tried to run my script and got an indentation error on Line 758 of data_types.py under CIP. I went and looked and sure enough, there was an extra space on line 757 before super(). Deleted the space and everything ran correctly. 

I uninstalled and recloned from github to see if I could recreate the problem and it wasn't there.

Anyway, where does configure_default_logger put the log file? I can't find it to save my life.

**Reply by ottowayi** [reply]
Weird, that isn't in my copy.  Umm if you just specify a file name it will save it in whatever your current working directory is, if you want to put in it a specific directory use the full path, e.g. `'c:/tmp/pycomm3.log'`.  In my case, I clone the repo and then add a test file I run, so the log shows up right in the same directory.:

```
pycomm3/
|-- docs/
|-- examples/
|-- pycomm3/
|    |-- cip/
|    |--logix_driver.py
|    ...
| ...
|-- setup.py
|-- test.py  <- my script
|-- pycomm3.log <- the log file 
```

**Comment by Colt-H** [comment]
I don't have an EN2T in the lab like I thought I did, they've all been deployed. I have some newer CompactLogix processors to test extended forward open on, but all my terminal blocks have been robbed and I have to order some more. 

How would I go about using Tox to run unit tests? I just flashed a processor to V32 to try to run the full test set on it.

**Reply by ottowayi** [reply]
1 - import the `pycomm3.L5X` program into the PLC (in the `tests` dir in the repo)
2 - install `tox` (`pip install tox`)
3 - open a terminal and `cd` into the pycomm3 repo
4 - run the command `tox -e user`

That should be all you need to do.  It will run all the offline tests and most of the online ones.  There are a few online tests that a specific to my PLC (matching serial numbers and other unique data) that this command will skip.

**Reply by JavierArmenta** [reply]
Hi im from reddit!. 
Right now i installed  tox and works ok. I ran the tests, offline is ok but online has errors. I suppose is because i don't set the ip address of the PLC. 
Can you help me with that ? where can i set that?  
I already did some simple write and read and works fine but i want to give you the log to help you. 
Thank you!

**Reply by ottowayi** [reply]
ohh yeah, I missed that part, sorry about that.  Before running the tox (step 4) - run a command to set the environment variable `PLCPATH`  Like: `set PLCPATH=192.168.1.236` (or if you're on Linux it's `export ...`)

**Comment by Colt-H** [comment]
I didn't want to put this under the Issues tab since it's getting into some very specific use cases that might be outside of the scope of testing for V1.0.0, and it's a good possibility I'm doing something wrong. 

When sending a generic message to get the mac address of this switch, it returns an array of 6 USINTs. I've tried every data type that would unpack correctly with no luck. When attempting to use any 64 bit data type (LINT, ULINT, LWORD) it returns an error `unpack requires a buffer of 8 bytes` from line 158 in data_types.py. I'm not sure if that's related to me trying to fit too much data in or what. I checked the struct.format for all the classes and it looks to be correct. All other related data types will just return a portion of the returned packet. I don't have any other way to test a 64 bit data type currently. 

Ideas?

**Reply by ottowayi** [reply]
So the L-data types are 8 bytes in size, but your array is only 6.  You can actually create array types very easy, just do `USINT[6]` and that should work.

**Reply by Colt-H** [reply]
I had no idea I could just do it that way.

I tried really hard last night to break generic messaging and this was the best I could come up with.

**Reply by ottowayi** [reply]
yeah sorry I really need to get this documentation written, but every time I try I get like 2 sentences in and then something else comes up.

**Reply by Colt-H** [reply]
No worries. I can take a stab at updating the docs for stuff I know how to use and know features of.

**Reply by ottowayi** [reply]
Thanks for the offer, but I'm going to do more than just an update.  I think they need to be restructured, right now the bulk of it is in a single page just for the LogixDriver.

**Reply by Colt-H** [reply]
Sounds good man. I definitely don't want to create more work by helping lol.
